### PR TITLE
Implement KHR_image_format_list validation

### PR DESCRIPTION
**Pull request by chrisforbes**

### CODE REVIEW

**Comment by chrisforbes** [comment]
Test case yet to come.

File: layers/buffer_validation.cpp [hunk 1]
```diff
@@ -3288,6 +3288,16 @@ bool ValidateImageBarrierSubresourceRange(const layer_data *device_data, const I
                                          subresourceRangeErrorCodes);
 }
 
+template<typename T>
+static T const * GetExtensionStruct(void const *base_struct, VkStructureType stype) {
```
Comment (tobine) [inline layers/buffer_validation.cpp]: Nice, should be able to expand use of this. May want to hoist to vk_layer_utils.

File: layers/buffer_validation.cpp [hunk 2]
```diff
@@ -3332,6 +3342,21 @@ bool PreCallValidateCreateImageView(layer_data *device_data, const VkImageViewCr
...
                                     validation_error_map[VALIDATION_ERROR_0ac007f4]);
                 }
+
+                else if (GetDeviceExtensions(device_data)->vk_khr_image_format_list) {
+                    auto format_list = GetExtensionStruct<VkImageFormatListCreateInfoKHR>(image_state->createInfo.ptr(),
+                                                                                          VK_STRUCTURE_TYPE_IMAGE_FORMAT_LIST_CREATE_INFO_KHR);
+                    if (format_list && format_list->viewFormatCount) {
+                        auto end = format_list->pViewFormats + format_list->viewFormatCount;
+                        if (std::find(format_list->pViewFormats, end, view_format) == end) {
+                            skip |= log_msg(report_data, VK_DEBUG_REPORT_ERROR_BIT_EXT, VK_DEBUG_REPORT_OBJECT_TYPE_UNKNOWN_EXT, 0,
```
Comment (tobine) [inline layers/buffer_validation.cpp]: I'd include the image as the object here.

**Comment by chrisforbes** [comment]
This is no good. The codegen for `safe_*` doesn't do the right thing with `pNext`. Will have to fix that first, and then this will be the right thing.

**Comment by mark-lunarg** [comment]
@jzulauf-lunarg, was that you that was taking a look at this pNext issue?

File: layers/buffer_validation.cpp (hunk 1 above)
Comment (jzulauf-lunarg) [reply layers/buffer_validation.cpp]: I've already written this template... 

 lvl_find_in_chain in vk_typemap_helper.h, which doesn't need an stype arge, but pulls the stype from a codegen'd traits for all T with pNext.

The traits don't exist for the safe structs, but could, or an override with an explicit sType could be added, or a SFINAE version could be written to map from the "safe" to "Vk" version through another autogenerated traits.

File: layers/buffer_validation.cpp [hunk 3]
```diff
@@ -3332,6 +3342,21 @@ bool PreCallValidateCreateImageView(layer_data *device_data, const VkImageViewCr
                                     __LINE__, VALIDATION_ERROR_0ac007f4, "IMAGE", "%s %s", ss.str().c_str(),
                                     validation_error_map[VALIDATION_ERROR_0ac007f4]);
                 }
+
+                else if (GetDeviceExtensions(device_data)->vk_khr_image_format_list) {
+                    auto format_list = GetExtensionStruct<VkImageFormatListCreateInfoKHR>(image_state->createInfo.ptr(),
+                                                                                          VK_STRUCTURE_TYPE_IMAGE_FORMAT_LIST_CREATE_INFO_KHR);
```
Comment (jzulauf-lunarg) [inline layers/buffer_validation.cpp]: The duplication of the type/stype isn't needed with lvl_find_in_chain.

**Pull request closed without merging**